Add virtual gamepad for touchscreen devices

StephenCWills opened the pull request:
This is an extremely rough draft of a virtual gamepad for touchscreen devices.

The buttons are circles, but are drawn as half-transparent rectangles because it was easy. Try to avoid touching the corners of the rectangles because they're not inside the button. Also, I found I couldn't see them at all when I went down to the dungeon, but it works okay when the player is in town.

The d-pad is on the far left. The four buttons on the far right simulate the A/B/X/Y buttons on a physical gamepad, however the controls were simplified quite a bit. For instance, the bottom button doesn't open the speedbook and basically only acts as a cancel/close button.

Rather than piling a ton of buttons onto the screen, I tried to compensate for the lack of buttons and simplified controls with some basic touchscreen interactions. You can open menus by tapping on the corresponding buttons in the bottom panel. You can also tap on options in menus to select them. In most other places, tapping simulates a mouse movement so you can view details in the info panel or select things to interact with.

On Windows at least, the OS would hide my pointer in response to touchscreen interactions. I recommend turning off hardware cursor when playing with touch controls.

I wanted to get straight to experimentation so I didn't try to implement a toggle for the virtual gamepad. If you play on this branch, the half-transparent rectangles will always be visible, and there is no way to hide them. Furthermore, the d-pad and buttons are in fixed positions relative to the screen's edge, and they will cover up parts of the bottom panel unless you play widescreen.

I completely got rid of the old touch controls. I also haven't tried to discriminate between touch devices on systems with multiple touchscreens so the front/back panel behavior for the Vita is just completely missing from this branch.

---

TODO list:

- [x] Replace half-transparent rectangles with proper graphics.
- [ ] Develop logic to hide/show the gamepad as appropriate based on user interaction.
- [ ] Enable user to adjust the position and size of gamepad controls.
- [ ] Implement some way of handling touch device selection for systems with multiple touchscreens
- [x] Maybe restore the Vita back panel functionality.
- [ ] Evaluate the controls to determine what should be kept and what should be changed.
- [ ] Review the code to ensure I'm not making a big mess. 😅

qndel commented:
Pics? : p

AJenbo commented:
Extremely exciting 🙏

StephenCWills commented:
> Pics? : p

Pics are uninteresting since there are no graphics yet:
![Screenshot_20210814-093805](https://user-images.githubusercontent.com/9203145/129448108-fd0debc1-0dbd-43e1-9c33-bac2ae9510ee.png)

AJenbo commented:
Since the controller can be contextual it makes sens that it would need less buttons then a real one.

Do you render the UI on the window surface or on the game surface?

StephenCWills commented:
> Do you render the UI on the window surface or on the game surface?

I'm pretty sure it's the game surface, since it's the same place where everything else gets drawn in-game (`DrawAndBlit()` in scrollrt).

AJenbo commented:
Gave it a quick spin on my phone. I have some early feedback:

- Controls are to small and to low on the screen
- Pick a different color then black, it's impossible to see

I realize this is pretty early for feedback like this, but I actually though there was something broken at first, but it was just me not hitting the controls right :)

> I'm pretty sure it's the game surface, since it's the same place where everything else gets drawn in-game (`DrawAndBlit()` in scrollrt).

Ok, we should probably consider doing it to the window surface instead so that it is scaled relatively to the device and not the game, but we can look in to that later.

AJenbo commented on Source/controls/touch/renderers.cpp:
```diff
@@ -0,0 +1,52 @@
+	int y = center.y - radius;
+	int width = diameter;
+	int height = diameter;
+	DrawHalfTransparentRectTo(out, x, y, width, height);
```
`SDL_FillRect(SDL_GetWindowSurface, SDL_MapRGBA)`

AJenbo replied:
So, looks like we can't draw to `SDL_GetWindowSurface` when hardware rendering is going on, but we should be able to draw to `GetOutputSurface()` which gives us nice 32bit support for the rendering.

glebm replied:
>  gives us nice 32bit support for the rendering

Only if the output surface is 32-bit (not the case on some platforms). Perhaps now it's best to stick to 8-bit.

AJenbo replied:
I assumed that anything wtih multi touch display also has 32bit capability. Is that not right?

glebm replied:
Running in 8-bit mode can still be beneficial for performance and battery life.
There are also things like HP TouchPad that have fewer bits for color.

StephenCWills commented:
I switched to green rectangles to demonstrate the **limitless power of the output surface** 😜. The controls should now be clearly visible in the dungeon, if not downright obtrusive.

glebm commented on Source/control.cpp:
```diff
@@ -1073,6 +1073,19 @@ void DoSpeedBook()
+{
+	for (int i = 0; i < PanelButtonIndex; i++) {
+		PanelButtons[i] = false;
+		drawbtnflag = true;
```
Move outside the loop?

glebm commented on Source/controls/controller_motion.cpp:
```diff
@@ -154,15 +155,15 @@ AxisDirection GetLeftStickOrDpadDirection(bool allowDpad)
 	AxisDirection result { AxisDirectionX_NONE, AxisDirectionY_NONE };
 
-	if (stickY >= 0.5 || (allowDpad && IsControllerButtonPressed(ControllerButton_BUTTON_DPAD_UP))) {
+	if (stickY >= 0.5 || (allowDpad && IsControllerButtonPressed(ControllerButton_BUTTON_DPAD_UP)) || virtualGamepad.IsControllerButtonPressed(ControllerButton_BUTTON_DPAD_UP)) {
```
Perhaps the virtual gamepad check should be moved into `IsControllerButtonPressed`?

StephenCWills replied:
Perhaps, but that would cause the virtual d-pad to be disabled along with the physical d-pad via the INI setting, which doesn't make much sense to me.

StephenCWills commented:
The latest build adds PNG support and incorporates graphics from BillieJoe for the d-pad and buttons. Some of the latest graphics he's provided have not yet been incorporated.

I did expect this version to fail on some platforms, but I've been caught by surprise to see the actual list. However, it's enough to see the Android build succeeding.

When testing this build, in order to see the graphics, you will need to copy the PNG files from `Packaging/resources/ui_art` in the source folder to `Android/data/org.diasurgical.devilutionx/files/ui_art` on the phone.
https://github.com/StephenCWills/devilutionX/tree/virtual-gamepad/Packaging/resources/ui_art

FYI, if you build this yourself on Windows or Linux, the `ui_art` folder should be automatically copied to the build folder.

StephenCWills commented:
Also, for @qndel, and anyone else who's interested, a screenshot from my phone.

![image](https://user-images.githubusercontent.com/9203145/129810197-e4ed42b1-9980-4639-81d0-b78ba1abbf66.png)

StephenCWills commented:
I fixed various CMake warnings and errors, and I got PNG support to build on a few more platforms. I'm not too sure what's going on with MacOSX, and I haven't touched the Windows builds just yet, but it looks like everything else is in good shape.

AJenbo commented:
It might soon be time to review this is suppose :)

glebm commented:
Please run `optipng -o9` on all the png files

glebm commented:
These graphics look 8-bit, perhaps we can stick to 8-bit for now and simply use the PCX format for them? What do you think? Don't want to bikeshed this too much though.

StephenCWills commented:
> These graphics look 8-bit, perhaps we can stick to 8-bit for now and simply use the PCX format for them? What do you think? Don't want to bikeshed this too much though.

I'm really not picky on what format we use, but PNG support is something we wanted to add to the project anyway. This provides a good use-case for loading PNG images. Now that I've got it working, what would be the justification for using PCX instead?

AJenbo commented:
PCX only supports binary (masked) transparency. but PNG lets us have 8bit transparency which can be a pretty big benefit for floating elements. The graphics do also use colors and gradients that wouldn't be available in 8bit.

![image](https://user-images.githubusercontent.com/204594/130664424-331622da-e6a0-4b58-b9ba-0e88a1540d6b.png)

I think we can look at creating an 8bit fallback path if we find a multi touch device that can't handle 20fps in full color mode :)

StephenCWills commented on docs/building.md:
````diff
@@ -133,19 +132,15 @@ Make sure to install the `C++ CMake tools for Windows` component for Visual Stud
+   vcpkg install fmt:x86-windows sdl2:x86-windows sdl2-ttf:x86-windows libsodium:x86-windows libpng:x86-windows gtest:x86-windows
    ```
 
-* **Manually**
-1. Download and place the MSVC Development Libraries of [SDL2](https://www.libsdl.org/download-2.0.php), [SDL2_ttf](https://www.libsdl.org/projects/SDL_ttf/) and [Libsodium](https://github.com/jedisct1/libsodium/releases) in `%USERPROFILE%\AppData\Local\Microsoft\WindowsApps\`.
-2. If dependencies are not found or you wish to place them in other location - configure required path variables in _"Manage Configurations..."_ dialog inside Visual Studio or in _cmake-gui_.
-
````
I feel like that video really needs some audio to help frame each step of the process. Like mentioning what webpages you're visiting and what each of the steps are meant to accomplish.

AJenbo replied:
Still not very used to recording my voice, but I might give it a shot, net time I install a Windows VM :D

Do you think the video is better then nothing, or should we just keep it as text only until someone do somethting better?

StephenCWills replied:
It's hard for me to say since I don't really use video tutorials myself. I feel like this video doesn't do much besides following the steps on the text guide so it may not be that useful for most. However, for anyone who has never used Visual Studio or really any IDE before, it might be useful to have this as a visual guide to verify that they're pressing the right buttons and getting the same feedback.

AJenbo replied:
Yeah visually verifying that they where pressing all the required buttons was the intent of the video. But I think I might be best to just have it handy when ever someone comes to the chat saying they couldn't get it to work, and then add a nicer once once it is available.

StephenCWills commented on docs/building.md:
````diff
@@ -41,7 +41,7 @@ cmake --build . -j $(sysctl -n hw.physicalcpu)
 ### Installing dependencies
 ```
-pkg install cmake sdl2_ttf libsodium
+pkg install cmake sdl2_ttf libsodium libpng
````
https://cgit.freebsd.org/ports/tree/graphics/png
```suggestion
pkg install cmake sdl2_ttf libsodium png
```

StephenCWills commented on docs/building.md:
````diff
@@ -54,7 +54,7 @@ cmake --build . -j $(sysctl -n hw.ncpu)
 ### Installing dependencies
 ```
-pkgin install cmake SDL2_ttf libsodium
+pkgin install cmake SDL2_ttf libsodium libpng
````
https://ftp.netbsd.org/pub/pkgsrc/current/pkgsrc/graphics/png/index.html
```suggestion
pkgin install cmake SDL2_ttf libsodium png
```

StephenCWills commented on docs/building.md:
````diff
@@ -68,7 +68,7 @@ cmake --build . -j $(sysctl -n hw.ncpu)
 ### Installing dependencies
 ```
-pkg_add cmake sdl2-ttf libsodium gmake
+pkg_add cmake sdl2-ttf libsodium libpng gmake
````
https://openports.se/graphics/png
```suggestion
pkg_add cmake sdl2-ttf libsodium png gmake
```

StephenCWills commented on docs/building.md:
````diff
@@ -236,19 +231,19 @@ cd build
 ### Installing dependencies on 32 bit Haiku
 ```
-pkgman install cmake_x86 devel:libsdl2_x86 devel:libsdl2_ttf_x86 devel:libsodium_x86
+pkgman install cmake_x86 devel:libsdl2_x86 devel:libsdl2_ttf_x86 devel:libsodium_x86 devel:libpng_x86
````
https://depot.haiku-os.org/libpng16_x86_devel
```suggestion
pkgman install cmake_x86 devel:libsdl2_x86 devel:libsdl2_ttf_x86 devel:libsodium_x86 devel:libpng16_x86
```

StephenCWills commented on docs/building.md:
````diff
@@ -236,19 +231,19 @@ cd build
 ### Installing dependencies on 64 bit Haiku
 ```
-pkgman install cmake devel:libsdl2 devel:libsdl2_ttf devel:libsodium
+pkgman install cmake devel:libsdl2 devel:libsdl2_ttf devel:libsodium devel:libpng
````
https://depot.haiku-os.org/libpng16_devel
```suggestion
pkgman install cmake devel:libsdl2 devel:libsdl2_ttf devel:libsodium devel:libpng16
```

StephenCWills reviewed (commented):
It looks like all the rest of the unverified package names were wrong.

AJenbo commented:
Dang :D

StephenCWills commented:
The commit history for this was getting a bit out of hand so I cleaned it up and fixed a couple issues along the way. I moved the precompiler checks around and added some conditional compilation so that the virtual gamepad code doesn't get compiled in on platforms that don't use it. I also added a precompiler variable called `VIRTUAL_GAMEPAD` to make it a bit easier for me to test on my touchscreen laptop.

This cleanup should make it significantly easier to review as well.

AJenbo commented on android-project/CMake/android_defs.cmake:
```diff
@@ -15,3 +17,6 @@ set(TTF_FONT_DIR \"\")
   COPY "${DevilutionX_SOURCE_DIR}/Packaging/resources/CharisSILB.ttf"
   DESTINATION "${DevilutionX_SOURCE_DIR}/android-project/app/src/main/assets")
+
+file(GLOB VirtualGamepadArt "${DevilutionX_SOURCE_DIR}/Packaging/resources/ui_art/*")
+file(COPY ${VirtualGamepadArt} DESTINATION "${DevilutionX_SOURCE_DIR}/android-project/app/src/main/assets/ui_art")
```
This feels like it soft of kind of depends on https://github.com/diasurgical/devilutionX/issues/2488

StephenCWills replied:
Yeah, and the code to load these assets is actually using `SFileOpenFile()`. Having a build-time option to include these assets in the MPQ would be nice.

StephenCWills commented on Source/controls/touch/renderers.h:
```diff
@@ -0,0 +1,156 @@
+public:
+	VirtualPadButtonRenderer(VirtualPadButton *virtualPadButton)
+	    : virtualPadButton(virtualPadButton)
+	    , attackSurface(nullptr)
+	    , pressedAttackSurface(nullptr)
+	    , talkSurface(nullptr)
+	    , pressedTalkSurface(nullptr)
+	    , itemSurface(nullptr)
+	    , pressedItemSurface(nullptr)
+	    , objectSurface(nullptr)
+	    , pressedObjectSurface(nullptr)
+	    , castSurface(nullptr)
+	    , pressedCastSurface(nullptr)
+	    , cancelSurface(nullptr)
+	    , pressedCancelSurface(nullptr)
+	    , blankSurface(nullptr)
+	    , pressedBlankSurface(nullptr)
```
I thought I took these out. They're completely unnecessary.
```suggestion
```

AJenbo replied:
As long as you don't rebase I think it will be fine :)

StephenCWills commented on Source/controls/touch/renderers.h:
```diff
@@ -0,0 +1,156 @@
+public:
+	VirtualDirectionPadRenderer(VirtualDirectionPad *virtualDirectionPad)
+	    : virtualDirectionPad(virtualDirectionPad)
+	    , padSurface(nullptr)
+	    , knobSurface(nullptr)
```
```suggestion
```

StephenCWills reviewed (commented):
@AJenbo I didn't want to change anything while you're in the middle of your review, but I did notice these when I was poking around just now.

AJenbo commented on Source/miniwin/misc_msg.cpp:
```diff
@@ -311,7 +315,22 @@ bool FetchMessage_Real(tagMSG *lpMsg)
 	}
 
-#ifndef USE_SDL1
+#if !defined(USE_SDL1) && !defined(__vita__)
+	if (!movie_playing) {
+		if (IsAnyOf(e.type, SDL_MOUSEBUTTONDOWN, SDL_MOUSEBUTTONUP) && e.button.which == SDL_TOUCH_MOUSEID)
+			return true;
+		if (e.type == SDL_MOUSEMOTION && e.motion.which == SDL_TOUCH_MOUSEID)
+			return true;
+		if (e.type == SDL_MOUSEWHEEL && e.wheel.which == SDL_TOUCH_MOUSEID)
+			return true;
+	}
+#endif
```
I'm a little unsure what this is handeling, could you add a comment for it?

My reading is that it ignores all touch input, but then I don't understand what the `HandleTouchEvent(e);` below is supposed to do ...

AJenbo replied:
Thanks, I forgot about the build in virtual mouse events in SDL :)

AJenbo commented on Source/control.cpp:
```diff
@@ -1065,6 +1065,14 @@ void DoSpeedBook()
 	SetCursorPos({ x, y });
 }
 
+void ClearPanBtn()
```
Seams like this could replace line 936-939 in `InitControlPan()`, it's a bit unclear what the state of `drawbtnflag` would be at that point, probably `true`, but I doubt it matters.

AJenbo replied:
Thanks, I forgot about the build in virtual mouse events in SDL :)

StephenCWills replied:
The call stack there is `ShowProgress()` -> `LoadGame()` -> `LoadGameLevel()` -> `InitControlPan()`. In other words, it's called while the load screen with the progress bar is showing so `drawbtnflag` should indeed be irrelevant.

AJenbo commented on Source/controls/touch/event_handlers.cpp:
```diff
@@ -0,0 +1,259 @@
+Point ScaleToScreenCoordinates(float x, float y)
+{
+	return Point {
+		(int)round(x * gnScreenWidth),
+		(int)round(y * gnScreenHeight)
```
Do we know if boarders are subtracted from this by SDL, or will we need to take that in to account?

StephenCWills replied:
SDL seems to take care of it. Tested on both Pixel 4 and a touchscreen Windows laptop.

AJenbo commented on Source/controls/touch/gamepad.cpp:
```diff
@@ -0,0 +1,140 @@
+	int padButtonSize = round(1.1 * screenPixels / 10);
+	int padButtonSpacing = inputMargin / 3;
+
+#ifndef USE_SDL1
```
Is this file even loaded under SDL1?

StephenCWills replied:
It was before I overhauled the precompiler checks. Good catch.

AJenbo replied:
If you could fix this then I think merging would be a no brainer at this point :)

StephenCWills replied:
My latest commit fixed it. 👍

AJenbo commented on Source/controls/touch/renderers.cpp:
```diff
@@ -0,0 +1,218 @@
+SDL_Surface *VirtualPadButtonRenderer::GetCancelSurface()
+{
+	if (cancelSurface == nullptr)
+		cancelSurface.reset(LoadPNG("ui_art\\back.png"));
```
I would suggest we combine all the buttons that are drawn in the same style/size and have an enum for the various states that are represented by the contained frames (similar to other sprites/graphics/buttons).

StephenCWills replied:
Can you point me to an example of one of these other sprites/graphics/buttons to give me a better idea of what you mean?

AJenbo replied:
```
enum {
     blank;
     back;
     backDown;
     pickup;
     pickupDown;
};

LoadMaskedArt("ui_art\\button.png", button, 5);

DrawArt(button, position, pressed ? pickupDown : pickup);
```

licaon-kter commented:
@StephenCWills None of the buttons can "use" potion/scroll? I see that I can move them to/from inventory though.

AJenbo commented:
Yes, that is still being worked out

StephenCWills commented:
The top button will use potions and scrolls when the inventory is open, but there is currently no graphic for it.

AJenbo commented:
Version 1.3.0 interface will probably be fairly bare bones, but after that I think it would be a good time to look in to how to add features like this.
